### PR TITLE
Added toast for ProblemResponse errors

### PR DESCRIPTION
**Pull request by Alexamakans**
- \[x] I've added a new note in the `CHANGELOG.md` file, according to docs:
  https://iver-wharf.github.io/#/development/changelogs/writing-changelogs

## Summary

- Added `ProblemResponse` handling to `GlobalErrorHandler`.
- Added `ProblemResponse` toast to `Notification` component.
- Import project form is now editable and resubmittable even after it errors.
  - Previously you had to refresh the page to be able to try importing again if the import failed for any reason.

`ProblemResponse` is referring to `Response` from [wharf-core](https://github.com/iver-wharf/wharf-core/blob/master/pkg/problem/response.go), which is [IETF RFC-7807](https://datatracker.ietf.org/doc/html/rfc7807) compliant.

## Motivation

More and better information in error messages, as well as a link to the documentation for that specific error.

### Old
old

### New
new
Clicking the `Read more` button takes you to the [documentation page](https://iver-wharf.github.io/#/prob/api-client/unexpected-write-error) for the `Unexpected API client write` error.

### CODE REVIEW

**Review by jilleJr — CHANGES_REQUESTED** [review]

File: src/app/shared/notification/notification.service.ts [hunk 1]
```diff
@@ -28,11 +29,31 @@ export class NotificationService {
...
   clear() {
     this.messageService.clear();
   }
 
   private showToast(message: string, type: string, isSticky = false) {
     this.messageService.add({severity: type, summary: type.toUpperCase(), detail: message, sticky: isSticky});
   }
+
+  private showProblemToast(problem: ProblemResponse, isSticky = false) {
+    const { title, detail, type, errors, status, instance } = problem;
```
[the remark this answers is not in the document]
Comment (Alexamakans) [reply src/app/shared/notification/notification.service.ts:45]: I like it because it makes my eyes dart left-and-right less when reading the object literal below it.
It also shows which properties from the problem object are used in a single line.

```typescript
const { title, detail, type, errors, status, instance } = problem;
this.messageService.add({
  summary: title,
  detail,
  severity: 'error',
  sticky: isSticky,
  data: {
    type,
    errors,
    status,
    instance,
    messageType: 'problem',
  },
});
```
vs.
```typescript
this.messageService.add({
  summary: problem.title,
  detail: problem.detail,
  severity: 'error',
  sticky: isSticky,
  data: {
    type: problem.type,
    errors: problem.errors,
    status: problem.status,
    instance: problem.instance,
    messageType: 'problem',
  },
});
```

File: src/app/shared/notification/notification.component.html [hunk 2]
```diff
@@ -24,6 +24,16 @@ <h3>{{message.summary}}</h3>
...
-
-
+<ng-template #problemToastContent let-message="message">
+  <h3>{{message.summary}}</h3>
+  <p id="message-details">{{message.detail}}</p>
+  <div id="message-errors" *ngFor="let error of message.data.errors">
+    <p>{{error}}</p>
+  </div>
+  <a pButton class="p-button-secondary"
+    *ngIf="message.data.type !== '' && message.data.type !== undefined && message.data.type !== 'about:blank'"
```
Comment (jilleJr) [inline src/app/shared/notification/notification.component.html:34]: nit: What if the `type` is `null`? :) A regular "truthyness" check ought to be enough

```suggestion
    *ngIf="message.data.type && message.data.type !== 'about:blank'"
```

File: src/app/shared/notification/notification.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jilleJr) [reply src/app/shared/notification/notification.service.ts:45]: Fare enough

**Review by jilleJr — APPROVED** [review]
LGTM! Make sure to check that code scanning warning. Perhaps it's simple to fix

File: src/app/shared/notification/notification.component.html (hunk 2 above)
Comment (Alexamakans) [reply src/app/shared/notification/notification.component.html:34]: Nice!
Just me not being JS-idiomatic from working in Go haha

File: src/app/shared/notification/notification.component.html [hunk 3]
```diff
@@ -24,6 +24,16 @@ <h3>{{message.summary}}</h3>
...
-
-
-
+<ng-template #problemToastContent let-message="message">
+  <h3>{{message.summary}}</h3>
+  <p id="message-details">{{message.detail}}</p>
+  <div id="message-errors" *ngFor="let error of message.data.errors">
+    <p>{{error}}</p>
+  </div>
+  <a pButton class="p-button-secondary"
```
Comment (jilleJr) [inline src/app/shared/notification/notification.component.html:33]: > Since it does not specify a link type, that page will be able to access the DOM of the origin page.
>
> https://github.com/iver-wharf/wharf-web/security/code-scanning/3?query=ref%3Arefs%2Fpull%2F54%2Fmerge+ref%3Arefs%2Fpull%2F54%2Fhead+ref%3Arefs%2Fheads%2Ffeature%2Fproblem-response-toast

Yo that sh*t's wild

We kind of know the URL will be safe, but it'll be nice to have that extra safe guard just in case

```suggestion
  <a pButton class="p-button-secondary"
     rel="noopener noreferrer"
```

**Review by jilleJr — APPROVED** [review]

File: src/app/shared/notification/notification.component.html (hunk 3 above)
Comment (Alexamakans) [reply src/app/shared/notification/notification.component.html:33]: Huh, that's wild indeed.

Didn't see it because it's medium, and I'm assuming that's why the check was green.
Now I know to always expand the checks! 👍

Comment (jilleJr) [reply src/app/shared/notification/notification.component.html:33]: It was visible through the "File changes" tab, as if it was yet another review comment

**Review by jilleJr — APPROVED** [review]

File: src/scss/base/reset/p-toast.scss [hunk 4]
```diff
@@ -23,5 +23,16 @@ body .p-toast .p-toast-message {
...
+
+    a,
+    button {
+      margin-right: 0.8em;
+    }
+
+    a:hover {
+      text-decoration: none;
+      background: #546e7a;
+      color: #fff;
```
Comment (fredx30) [inline src/scss/base/reset/p-toast.scss:35]: This would probably be part of a larger css rework but usually it would be good to link magic numbers such as color to scss theme variables such that the theme can be centrally controlled.

File: src/scss/base/reset/p-toast.scss [hunk 5]
```diff
@@ -23,5 +23,16 @@ body .p-toast .p-toast-message {
 
   &.p-toast-message-error {
     background-color: $wharf_status_red !important;
+
+    a,
+    button {
+      margin-right: 0.8em;
```
Comment (fredx30) [inline src/scss/base/reset/p-toast.scss:29]: I usually preffer rem over em as its typically easier to predict.
```suggestion
      margin-right: 0.8rem;
```

**Review by fredx30 — APPROVED** [review]

File: projects/api-client/src/api/artifact.service.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@
  * Wharf main API
  * Wharf backend API that manages data storage for projects, builds, providers, etc.
  *
- * OpenAPI spec version: v4.1.0
+ * OpenAPI spec version: local docker-compose
```
Comment (jilleJr) [inline projects/api-client/src/api/artifact.service.ts:5]: Should only building using the latest *released* version of the APIs

Suggest reverting this commit. Why did you need it regenerated anyway?

**Review by jilleJr — CHANGES_REQUESTED** [review]

File: projects/api-client/src/api/artifact.service.ts (hunk 6 above)
Comment (Alexamakans) [reply projects/api-client/src/api/artifact.service.ts:5]: Classifying this as an oops from having worked on the test results display.
I will regenerate with the proper versions, to make sure it's up-to-date.

Comment (jilleJr) [reply projects/api-client/src/api/artifact.service.ts:5]: I think regenerating these are out of scope for this PR though

Comment (Alexamakans) [reply projects/api-client/src/api/artifact.service.ts:5]: I see.
Will just revert then.

File: src/scss/base/reset/p-toast.scss (hunk 4 above)
Comment (Alexamakans) [reply src/scss/base/reset/p-toast.scss:35]: Good point.
Extracting these to `variables.scss` before they get forgotten is probably the best way to go, what do you think?

Comment (fredx30) [reply src/scss/base/reset/p-toast.scss:35]: The root scss files are specified here- https://github.com/iver-wharf/wharf-web/blob/df795e9358f8cba66f27878c7668269e0a52ecd7/angular.json#L29-L32

I think the natural thing to do would be to link the root file at `%repo_root%/styles.scss` for style that we intend to keep global. Then i would define specific themes in the `%repo_root%/scss/**/*` and use scss import declarations to import them whenever nessecary. Note the global file could also import specific or default themes from the `%repo_root%/scss/**/*`.This should probably be up for discussion though.

**Review by fredx30 — APPROVED** [review]
LGTM!

File: src/app/shared/error/global-error-handler.ts [hunk 7]
```diff
@@ -2,23 +2,42 @@ import { NotificationService } from './../notification/notification.service';
...
+    const { detail, errors, instance, status, title, type } = error;
+    return {
+      detail,
+      errors,
+      instance,
+      status,
+      title,
+      type,
+    };
+  }
```
Comment (jilleJr) [inline src/app/shared/error/global-error-handler.ts:42]: nit: Unconventional term `make` in TS code. Suggest perhaps `createProblem` or `createProblemFromError` to be more descriptive

**Review by jilleJr — APPROVED** [review]

File: src/app/providers/provider-form/azuredevops/azuredevops.component.ts [hunk 8]
```diff
@@ -38,8 +38,9 @@ export class AzureDevOpsComponent {
           this.providersService.triggerCloseForm(this.providerForm);
         },
         err => {
-          this.globalErrorHandler.handleError(err);
           console.log(err);
+          this.globalErrorHandler.handleError(err);
+          this.providerForm.enable();
         },
         () => {
           this.providerForm.enable();
```
[the remark this answers is not in the document]
Comment (Alexamakans) [reply src/app/providers/provider-form/azuredevops/azuredevops.component.ts]: Changed to the second version.

Also changed to use an object as argument to `subscribe`, as passing raw callbacks is planned for removal in RxJS v8.
https://rxjs.dev/deprecations/subscribe-arguments
